### PR TITLE
Adding a fixture to set scheduler to slower speeds and revert it back.

### PR DESCRIPTION
**Pull request by rraghav-cisco**
### Description of PR

Summary:
Fixes the flakiness of DWRR testcase. The PR adds a new fixture that slows down the scheduler without changing the underlying algorithm. This allows the dWRR test to pass consitently.

### Type of change

- [ ] Bug fix
- [X] Testbed and Framework(new/improvement)
- [ ] Test case(new/improvement)

### Back port request
- [ ] 202012
- [ ] 202205
- [ ] 202305
- [ ] 202311
- [X] 202405

### CODE REVIEW

**Comment by rraghav-cisco** [comment]
Result from my run:
```
=========================================================================================================================== PASSES ===========================================================================================================================
______________________________________________________________________________________________________ TestQosSai.testQosSaiDwrr[single_dut_multi_asic] ______________________________________________________________________________________________________
---------------------------------------------------- generated xml file: /run_logs/dwrr/2024-11-24-04-39-38/1/L-400g-LL-masic-gb/qos/test_qos_sai.py::TestQosSai::testQosSaiDwrr_2024-11-24-04-39-38.xml -----------------------------------------------------
INFO:root:Can not get Allure report URL. Please check logs
------------------------------------------------------------------------------------------------------------------- live log sessionfinish -------------------------------------------------------------------------------------------------------------------
04:56:32 __init__.pytest_terminal_summary         L0067 INFO   | Can not get Allure report URL. Please check logs
================================================================================================================== short test summary info ===================================================================================================================
PASSED qos/test_qos_sai.py::TestQosSai::testQosSaiDwrr[single_dut_multi_asic]
========================================================================================================= 1 passed, 1 warning in 1012.38s (0:16:52) ==========================================================================================================
sonic@202405-qos-sonic-mgmt-prod:/data/tests$
```

**Comment by rraghav-cisco** [comment]
Finally found a way to repeat tests:
```
====================================================================================================================== warnings summary ======================================================================================================================
../../usr/local/lib/python3.8/dist-packages/paramiko/transport.py:236
  /usr/local/lib/python3.8/dist-packages/paramiko/transport.py:236: CryptographyDeprecationWarning: Blowfish has been deprecated
    "class": algorithms.Blowfish,

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
=========================================================================================================================== PASSES ===========================================================================================================================
___________________________________________________________________________________________________ TestQosSai.testQosSaiDwrr[single_dut_multi_asic-1-10] ____________________________________________________________________________________________________
___________________________________________________________________________________________________ TestQosSai.testQosSaiDwrr[single_dut_multi_asic-2-10] ____________________________________________________________________________________________________
___________________________________________________________________________________________________ TestQosSai.testQosSaiDwrr[single_dut_multi_asic-3-10] ____________________________________________________________________________________________________
___________________________________________________________________________________________________ TestQosSai.testQosSaiDwrr[single_dut_multi_asic-4-10] ____________________________________________________________________________________________________
___________________________________________________________________________________________________ TestQosSai.testQosSaiDwrr[single_dut_multi_asic-5-10] ____________________________________________________________________________________________________
___________________________________________________________________________________________________ TestQosSai.testQosSaiDwrr[single_dut_multi_asic-6-10] ____________________________________________________________________________________________________
___________________________________________________________________________________________________ TestQosSai.testQosSaiDwrr[single_dut_multi_asic-7-10] ____________________________________________________________________________________________________
___________________________________________________________________________________________________ TestQosSai.testQosSaiDwrr[single_dut_multi_asic-8-10] ____________________________________________________________________________________________________
___________________________________________________________________________________________________ TestQosSai.testQosSaiDwrr[single_dut_multi_asic-9-10] ____________________________________________________________________________________________________
___________________________________________________________________________________________________ TestQosSai.testQosSaiDwrr[single_dut_multi_asic-10-10] ___________________________________________________________________________________________________
---------------------------------------------------- generated xml file: /run_logs/dwrr/2024-11-24-19-05-56/1/L-400g-LL-masic-gb/qos/test_qos_sai.py::TestQosSai::testQosSaiDwrr_2024-11-24-19-05-56.xml -----------------------------------------------------
INFO:root:Can not get Allure report URL. Please check logs
------------------------------------------------------------------------------------------------------------------- live log sessionfinish -------------------------------------------------------------------------------------------------------------------
19:35:30 __init__.pytest_terminal_summary         L0067 INFO   | Can not get Allure report URL. Please check logs
================================================================================================================== short test summary info ===================================================================================================================
PASSED qos/test_qos_sai.py::TestQosSai::testQosSaiDwrr[single_dut_multi_asic-1-10]
PASSED qos/test_qos_sai.py::TestQosSai::testQosSaiDwrr[single_dut_multi_asic-2-10]
PASSED qos/test_qos_sai.py::TestQosSai::testQosSaiDwrr[single_dut_multi_asic-3-10]
PASSED qos/test_qos_sai.py::TestQosSai::testQosSaiDwrr[single_dut_multi_asic-4-10]
PASSED qos/test_qos_sai.py::TestQosSai::testQosSaiDwrr[single_dut_multi_asic-5-10]
PASSED qos/test_qos_sai.py::TestQosSai::testQosSaiDwrr[single_dut_multi_asic-6-10]
PASSED qos/test_qos_sai.py::TestQosSai::testQosSaiDwrr[single_dut_multi_asic-7-10]
PASSED qos/test_qos_sai.py::TestQosSai::testQosSaiDwrr[single_dut_multi_asic-8-10]
PASSED qos/test_qos_sai.py::TestQosSai::testQosSaiDwrr[single_dut_multi_asic-9-10]
PASSED qos/test_qos_sai.py::TestQosSai::testQosSaiDwrr[single_dut_multi_asic-10-10]
========================================================================================================= 10 passed, 1 warning in 1773.04s (0:29:33) =========================================================================================================
sonic@202405-qos-sonic-mgmt-prod:/data/tests$ 
```

**Comment by rraghav-cisco** [comment]
100G/S - 100G/S:
```
INFO:root:Can not get Allure report URL. Please check logs
------------------------------------------------------------------------------------------------------------------- live log sessionfinish -------------------------------------------------------------------------------------------------------------------
01:56:02 __init__.pytest_terminal_summary         L0067 INFO   | Can not get Allure report URL. Please check logs
================================================================================================================== short test summary info ===================================================================================================================
PASSED qos/test_qos_sai.py::TestQosSai::testQosSaiDwrr[single_dut_multi_asic-1-10]
PASSED qos/test_qos_sai.py::TestQosSai::testQosSaiDwrrWeightChange[single_dut_multi_asic-1-10]
PASSED qos/test_qos_sai.py::TestQosSai::testQosSaiDwrr[single_dut_multi_asic-2-10]
PASSED qos/test_qos_sai.py::TestQosSai::testQosSaiDwrrWeightChange[single_dut_multi_asic-2-10]
PASSED qos/test_qos_sai.py::TestQosSai::testQosSaiDwrr[single_dut_multi_asic-3-10]
PASSED qos/test_qos_sai.py::TestQosSai::testQosSaiDwrrWeightChange[single_dut_multi_asic-3-10]
PASSED qos/test_qos_sai.py::TestQosSai::testQosSaiDwrr[single_dut_multi_asic-4-10]
PASSED qos/test_qos_sai.py::TestQosSai::testQosSaiDwrrWeightChange[single_dut_multi_asic-4-10]
PASSED qos/test_qos_sai.py::TestQosSai::testQosSaiDwrr[single_dut_multi_asic-5-10]
PASSED qos/test_qos_sai.py::TestQosSai::testQosSaiDwrrWeightChange[single_dut_multi_asic-5-10]
PASSED qos/test_qos_sai.py::TestQosSai::testQosSaiDwrr[single_dut_multi_asic-6-10]
PASSED qos/test_qos_sai.py::TestQosSai::testQosSaiDwrrWeightChange[single_dut_multi_asic-6-10]
PASSED qos/test_qos_sai.py::TestQosSai::testQosSaiDwrr[single_dut_multi_asic-7-10]
PASSED qos/test_qos_sai.py::TestQosSai::testQosSaiDwrrWeightChange[single_dut_multi_asic-7-10]
PASSED qos/test_qos_sai.py::TestQosSai::testQosSaiDwrr[single_dut_multi_asic-8-10]
PASSED qos/test_qos_sai.py::TestQosSai::testQosSaiDwrrWeightChange[single_dut_multi_asic-8-10]
PASSED qos/test_qos_sai.py::TestQosSai::testQosSaiDwrr[single_dut_multi_asic-9-10]
PASSED qos/test_qos_sai.py::TestQosSai::testQosSaiDwrrWeightChange[single_dut_multi_asic-9-10]
PASSED qos/test_qos_sai.py::TestQosSai::testQosSaiDwrr[single_dut_multi_asic-10-10]
PASSED qos/test_qos_sai.py::TestQosSai::testQosSaiDwrrWeightChange[single_dut_multi_asic-10-10]
ERROR qos/test_qos_sai.py::TestQosSai::testQosSaiDwrr[single_dut_multi_asic-1-10] - Failed: Processes "['analyze_logs--<MultiAsicSonicHost xx37-lc7>']" failed with exit code "1"
ERROR qos/test_qos_sai.py::TestQosSai::testQosSaiDwrrWeightChange[single_dut_multi_asic-1-10] - Failed: Processes "['analyze_logs--<MultiAsicSonicHost xx37-lc7>']" failed with exit code "1"
ERROR qos/test_qos_sai.py::TestQosSai::testQosSaiDwrr[single_dut_multi_asic-2-10] - Failed: Processes "['analyze_logs--<MultiAsicSonicHost xx37-lc7>']" failed with exit code "1"
ERROR qos/test_qos_sai.py::TestQosSai::testQosSaiDwrrWeightChange[single_dut_multi_asic-2-10] - Failed: Processes "['analyze_logs--<MultiAsicSonicHost xx37-lc7>']" failed with exit code "1"
ERROR qos/test_qos_sai.py::TestQosSai::testQosSaiDwrr[single_dut_multi_asic-3-10] - Failed: Processes "['analyze_logs--<MultiAsicSonicHost xx37-lc7>']" failed with exit code "1"
ERROR qos/test_qos_sai.py::TestQosSai::testQosSaiDwrrWeightChange[single_dut_multi_asic-3-10] - Failed: Processes "['analyze_logs--<MultiAsicSonicHost xx37-lc7>']" failed with exit code "1"
ERROR qos/test_qos_sai.py::TestQosSai::testQosSaiDwrr[single_dut_multi_asic-4-10] - Failed: Processes "['analyze_logs--<MultiAsicSonicHost xx37-lc7>']" failed with exit code "1"
ERROR qos/test_qos_sai.py::TestQosSai::testQosSaiDwrrWeightChange[single_dut_multi_asic-4-10] - Failed: Processes "['analyze_logs--<MultiAsicSonicHost xx37-lc7>']" failed with exit code "1"
ERROR qos/test_qos_sai.py::TestQosSai::testQosSaiDwrr[single_dut_multi_asic-5-10] - Failed: Processes "['analyze_logs--<MultiAsicSonicHost xx37-lc7>']" failed with exit code "1"
ERROR qos/test_qos_sai.py::TestQosSai::testQosSaiDwrrWeightChange[single_dut_multi_asic-5-10] - Failed: Processes "['analyze_logs--<MultiAsicSonicHost xx37-lc7>']" failed with exit code "1"
ERROR qos/test_qos_sai.py::TestQosSai::testQosSaiDwrr[single_dut_multi_asic-6-10] - Failed: Processes "['analyze_logs--<MultiAsicSonicHost xx37-lc7>']" failed with exit code "1"
ERROR qos/test_qos_sai.py::TestQosSai::testQosSaiDwrrWeightChange[single_dut_multi_asic-6-10] - Failed: Processes "['analyze_logs--<MultiAsicSonicHost xx37-lc7>']" failed with exit code "1"
ERROR qos/test_qos_sai.py::TestQosSai::testQosSaiDwrr[single_dut_multi_asic-7-10] - Failed: Processes "['analyze_logs--<MultiAsicSonicHost xx37-lc7>']" failed with exit code "1"
ERROR qos/test_qos_sai.py::TestQosSai::testQosSaiDwrrWeightChange[single_dut_multi_asic-7-10] - Failed: Processes "['analyze_logs--<MultiAsicSonicHost xx37-lc7>']" failed with exit code "1"
ERROR qos/test_qos_sai.py::TestQosSai::testQosSaiDwrr[single_dut_multi_asic-8-10] - Failed: Processes "['analyze_logs--<MultiAsicSonicHost xx37-lc7>']" failed with exit code "1"
ERROR qos/test_qos_sai.py::TestQosSai::testQosSaiDwrrWeightChange[single_dut_multi_asic-8-10] - Failed: Processes "['analyze_logs--<MultiAsicSonicHost xx37-lc7>']" failed with exit code "1"
ERROR qos/test_qos_sai.py::TestQosSai::testQosSaiDwrr[single_dut_multi_asic-9-10] - Failed: Processes "['analyze_logs--<MultiAsicSonicHost xx37-lc7>']" failed with exit code "1"
ERROR qos/test_qos_sai.py::TestQosSai::testQosSaiDwrrWeightChange[single_dut_multi_asic-9-10] - Failed: Processes "['analyze_logs--<MultiAsicSonicHost xx37-lc7>']" failed with exit code "1"
ERROR qos/test_qos_sai.py::TestQosSai::testQosSaiDwrr[single_dut_multi_asic-10-10] - Failed: Processes "['analyze_logs--<MultiAsicSonicHost xx37-lc7>']" failed with exit code "1"
ERROR qos/test_qos_sai.py::TestQosSai::testQosSaiDwrrWeightChange[single_dut_multi_asic-10-10] - Failed: Processes "['analyze_logs--<MultiAsicSonicHost xx37-lc7>']" failed with exit code "1"
=========================================================================================== 20 passed, 470 deselected, 1 warning, 20 errors in 2973.01s (0:49:33) ============================================================================================
sonic@202405-qos-sonic-mgmt-prod:/data/tests$ 
```

**Comment by rraghav-cisco** [comment]
100G/S - 100G/S - sAsic:
```
---------------------------------------------------------- generated xml file: /run_logs/dwrr-cir/2024-11-26-02-31-41/1/O-100g-SS-gb-sasic/qos/test_qos_sai.py::TestQosSai_2024-11-26-02-31-41.xml -----------------------------------------------------------
INFO:root:Can not get Allure report URL. Please check logs
------------------------------------------------------------------------------------------------------------------- live log sessionfinish -------------------------------------------------------------------------------------------------------------------
03:05:14 __init__.pytest_terminal_summary         L0067 INFO   | Can not get Allure report URL. Please check logs
================================================================================================================== short test summary info ===================================================================================================================
PASSED qos/test_qos_sai.py::TestQosSai::testQosSaiDwrr[single_asic-1-10]
PASSED qos/test_qos_sai.py::TestQosSai::testQosSaiDwrrWeightChange[single_asic-1-10]
PASSED qos/test_qos_sai.py::TestQosSai::testQosSaiDwrr[single_asic-2-10]
PASSED qos/test_qos_sai.py::TestQosSai::testQosSaiDwrrWeightChange[single_asic-2-10]
PASSED qos/test_qos_sai.py::TestQosSai::testQosSaiDwrr[single_asic-3-10]
PASSED qos/test_qos_sai.py::TestQosSai::testQosSaiDwrrWeightChange[single_asic-3-10]
PASSED qos/test_qos_sai.py::TestQosSai::testQosSaiDwrr[single_asic-4-10]
PASSED qos/test_qos_sai.py::TestQosSai::testQosSaiDwrrWeightChange[single_asic-4-10]
PASSED qos/test_qos_sai.py::TestQosSai::testQosSaiDwrr[single_asic-5-10]
PASSED qos/test_qos_sai.py::TestQosSai::testQosSaiDwrrWeightChange[single_asic-5-10]
PASSED qos/test_qos_sai.py::TestQosSai::testQosSaiDwrr[single_asic-6-10]
PASSED qos/test_qos_sai.py::TestQosSai::testQosSaiDwrrWeightChange[single_asic-6-10]
PASSED qos/test_qos_sai.py::TestQosSai::testQosSaiDwrr[single_asic-7-10]
PASSED qos/test_qos_sai.py::TestQosSai::testQosSaiDwrrWeightChange[single_asic-7-10]
PASSED qos/test_qos_sai.py::TestQosSai::testQosSaiDwrr[single_asic-8-10]
PASSED qos/test_qos_sai.py::TestQosSai::testQosSaiDwrrWeightChange[single_asic-8-10]
PASSED qos/test_qos_sai.py::TestQosSai::testQosSaiDwrr[single_asic-9-10]
PASSED qos/test_qos_sai.py::TestQosSai::testQosSaiDwrrWeightChange[single_asic-9-10]
PASSED qos/test_qos_sai.py::TestQosSai::testQosSaiDwrr[single_asic-10-10]
PASSED qos/test_qos_sai.py::TestQosSai::testQosSaiDwrrWeightChange[single_asic-10-10]
================================================================================================= 20 passed, 470 deselected, 1 warning in 2011.91s (0:33:31) =================================================================================================
sonic@202405-qos-sonic-mgmt-prod:/data/tests$ 
```

**Comment by rraghav-cisco** [comment]
I am not sure what the error means in the prechecks. @auspham , pls help.

File: tests/qos/qos_sai_base.py [hunk 1]
```diff
@@ -2624,3 +2624,60 @@ def change_lag_lacp_timer(self, duthosts, get_src_dst_asic_and_duts, tbinfo, nbr
...
+
+        # Set scheduler to 5 Gbps.
+        self.copy_and_run_set_cir_script_cisco_8000(
+            dut=dst_dut, port=dst_port, asic=dst_index, speed=5 * 1000 * 1000 * 1000)
+
+        yield
+
+        # Set scheduler back to original speed.
+        self.copy_and_run_set_cir_script_cisco_8000(
+            dut=dst_dut, port=dst_port, asic=dst_index, speed=int(1.1*port_speed))
```
Comment (sdszhang) [inline tests/qos/qos_sai_base.py:2683]: why we need to use 1.1*port_speed?

**Comment by zhixzhu** [comment]
@rraghav-cisco tx enable/disable impacts scheduler rate. Need to adjust sequence of steps as below:
1, tx disable, this set credit cir to 0
2, send packets
3, set_credit_cir to 5Gbps
4, check received packets
5, tx enable, this set credit cir to original value.

No need to set_credit_eir_or_pir().

File: tests/qos/qos_sai_base.py [hunk 2]
```diff
@@ -2624,3 +2624,60 @@ def change_lag_lacp_timer(self, duthosts, get_src_dst_asic_and_duts, tbinfo, nbr
...
+            match = re.match(pattern, output)
+            if not match:
+                raise RuntimeError(f"Couldn't find required interfaces out of the output:{output}")
+            interfaces = match.group(1).split(' ')
+
+        for intf in interfaces:
+            # Set scheduler to 5 Gbps.
+            self.copy_and_run_set_cir_script_cisco_8000(
+                dut=dst_dut,
+                port=intf,
```
Comment (sdszhang) [inline tests/qos/qos_sai_base.py:2678]: Got this error:

        for intf in interfaces:
            # Set scheduler to 5 Gbps.
>           self.copy_and_run_set_cir_script_cisco_8000(
                dut=dst_dut,
                port=intf,
                asic=dst_index,
                speed=5 * 1000 * 1000 * 1000)
E           TypeError: copy_and_run_set_cir_script_cisco_8000() got an unexpected keyword argument 'port'

Comment (rraghav-cisco) [reply tests/qos/qos_sai_base.py]: @sdszhang , I apologize. I missed the update to this function in my previous commit. I have fixed it now. Thanks.

**Review by sdszhang — APPROVED** [review]

**Comment by yejianquan** [comment]
@rraghav-cisco 
Please resolve the conflict, 
And I'm a little bit curious that, where do we set back the scheduler to original value?

**Comment by rraghav-cisco** [comment]
> @rraghav-cisco Please resolve the conflict, And I'm a little bit curious that, where do we set back the scheduler to original value?

@yejianquan : I will resolve the conflict. The set-back happens with tx_enable inside the ptf script. The function: sai_thrift_port_tx_enable() in the file: tests/saitests/py3/sai_qos_tests.py, which is run in PTF, basically reverts the scheduler to original state.

**Comment by rraghav-cisco** [comment]
PR for 202405: https://github.com/sonic-net/sonic-mgmt/pull/16199

**Comment by sdszhang** [comment]
202405 manual cherry pick PR #16199, merged.

**Pull request merged**